### ISSUE TITLE
Record Server-Timing metrics during GH workflows

### ISSUE THREAD

**Issue by mukeshpanchal27** [10up/wordpress-develop#4]
## Overview
The purpose of this task is to implement the process that performs the performance measurements during the GitHub workflow introduced in #6. This process will rely on the MU plugin from #7 to measure server timing.

## Requirements
- [ ] The CLI script at `/tests/performance/cli.js` is updated to include the logic needed to perform performance measurements
- [ ] Before performance measurements are run, import content from the [Theme Test data](https://github.com/WPTT/theme-test-data) ([raw](https://raw.githubusercontent.com/WPTT/theme-unit-test/master/themeunittestdata.wordpress.xml))
- [ ] Permalink structures should be set to `'/%year%/%monthnum%/%postname%/'`
- [ ] The "all-pages" menu from the test data should be set to the "primary" location, i.e. `wp menu location assign all-pages primary`.
- [ ] Measurements should be taken for both the Twenty Twenty One (classic), and Twenty Twenty Three (block theme) themes
- [ ] For each measurement, 10 requests should be made to the homepage, and the median value for the following metrics should be recorded from the Server-Timing API
	- [ ] Total Time
	- [ ] Before Template
	- [ ] Template
- [ ] The median measurements for each theme and metric should be logged out to the console during the workflow.

**Not in scope for this iteration**
- Measurement should be taken for the current commit as well as the previous commit for comparison (per: https://github.com/10up/wordpress-develop/issues/4#issuecomment-1416379325)

## Implementation notes

- Gutenberg has [a CLI script](https://github.com/WordPress/gutenberg/blob/trunk/bin/plugin/commands/performance.js) that is similar, which can be referenced
- The implentation shoud rely on either the e2e tools already in use by core, or make a simple cURL or http.request to collect server timing metrics
- The e2e-utils package includes several utilities for performing actions like [activating a theme](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-e2e-test-utils/#activatetheme) during test runs, otherwise WP CLI might be useful.

**Comment by adamsilverstein** [comment]
> Undecided (Run the home page)

I propose we start with the latest core theme home page with it's dummy content (eg https://wordpress.org/themes/twentytwentythree/) and the last (?) classic core theme w/demo content - eg https://wordpress.org/themes/twentytwentyone/

> Print the result on GH action.

What about sending the results to the dashboard?  Are you creating a separate issue for that?

**Comment by joemcgill** [comment]
We may be able to reference the process being used in the Gutenberg repo for measuring TTFB here. See this Slack Post for reference: https://wordpress.slack.com/archives/C02KGN5K076/p1674725621023759

**Comment by joemcgill** [comment]
This is currently blocked until we've gotten #1 sorted.

**Comment by joemcgill** [comment]
@adamsilverstein re:

> What about sending the results to the dashboard? Are you creating a separate issue for that?

I've created #5 for implementing logging to the dashboard. However, we may find that we can consolidate that work with this one during implementation if it's easier to do it all at once.

**Comment by felixarntz** [comment]
@mukeshpanchal27 As @joemcgill is saying, I think we should decide for the ACs here how at a high level we're going to do the measurement:
1. Either a simple request so that we get back the response headers including `Server-Timing`
2. Or a headless browser which also allows us to measure TTFB (and in the future any client-side metrics), and we could get the Server-Timing metrics via `performance.getEntries()[0].serverTiming`

The latter is definitely more forward-thinking, but at least in theory it is also a larger effort, and we should be conscious of that as we would like to get the MVP ready sooner than later. However, since Gutenberg already has option 2, we may just be able to copy a lot of their code, which makes option 2 potentially more appealing. Would it be possible to come up with an estimate on the work for either option? What are your thoughts on that?

**Comment by joemcgill** [comment]
I think a simple request to get header info is totally fine for this task in order to get a proof of concept more quickly. However, if it doesn't add too much time/complexity and we can reuse the approach the Gutenberg team are working on [here](https://github.com/WordPress/gutenberg/pull/47037/files), I think that would be better, because it will open up the opportunity for us to expand the core tests with additional performance metrics more quickly.

@mukeshpanchal27 if you could take a look at the approach in the Gutenberg repo and share your thoughts about the feasibility of implementing a similar process in this repo, I think that will help us decide and finalize the A/C here.

**Comment by mukeshpanchal27** [comment]
@joemcgill I looked into the Gutenberg approach, but I'm having trouble coming up with an extract idea because there are so many different scripts, and I've also tried to set up something similar in my GH, but it's never worked for me, and I'm stuck.

Because I have to check everything in GH, it is difficult to troubleshoot the error, so if you could assist me, that would be great.

**Comment by joemcgill** [comment]
The way the Gutenberg approach currently works is this:

First, the workflow is calling the following code on PRs (a similar task is run on push events, but this is easier to read):

```
            - name: Compare performance with trunk
              if: github.event_name == 'pull_request'
              run: ./bin/plugin/cli.js perf $GITHUB_SHA trunk --tests-branch $GITHUB_SHA
```

Which eventually passes a CLI command to the exported `runPerformanceTests()` function in  `/bin/plugin/commands/performance.js`. That function is doing all the heavy lifting by setting up test directories, cloning the different WP versions needed for their tests (not something we should need here), running the performance tests, and then formatting and recording the results to temporary text files.

Finally, this workflow has a step that logs the timing, which is similar to what we'll want to do in #5.

```
            - name: Publish performance results
              if: github.event_name == 'push'
              env:
                  COMMITTED_AT: ${{ steps.commit-timestamp.outputs.result }}
                  CODEHEALTH_PROJECT_TOKEN: ${{ secrets.CODEHEALTH_PROJECT_TOKEN }}
              run: ./bin/log-perormance-results.js $CODEHEALTH_PROJECT_TOKEN trunk $GITHUB_SHA debd225d007f4e441ceec80fbd6fa96653f94737 $COMMITTED_AT
```

WP Core doesn't include the same CLI infrastructure as the Gutenberg repo, so I think we'll want to add a new folder to the `/tests/` called `/performance/` which includes the JS that we'll want to run for recording the server timing. I think we can also simplify what that script does quite a bit for our purposes. While it would be nice to use the `test-e2e` package from `wp-scripts` and pass a custom spec, perhaps an initial approach could simply demonstrate sending a HTTP request to the homepage and record the header value. We'll want to send multiple requests in order to get the median value from several independent runs?

If you point me to a specific place in your own fork where you're seeing an error, I could probably help debug better. Also happy to help work up an initial approach if needed.

**Comment by mukeshpanchal27** [comment]
@joemcgill I've made some progress in spinning up a WP environment for two branches, but I'm stuck at [e2e testing for themes that fail](https://github.com/mukeshpanchal27/wordpress-develop/pull/16); could you please explain why this fails and how I can proceed?

**Comment by joemcgill** [comment]
@mukeshpanchal27 it looks to me like those tests are failing due to file changes occurring during the test run. You'll likely need to add a rule to the `.gitignore` file to make sure any files generated during tests are not picked up by git.

**Comment by joemcgill** [comment]
@mukeshpanchal27 I've updated the AC on this one to hopefully clarify the requirements a bit, based on conversations from #1. Let me know if you have any questions. One thing I'm still unclear about is the criteria that "Measurement should be taken for the current commit as well as the previous commit for comparison". @adamsilverstein or @felixarntz do either of you know if this is necessary?

**Comment by felixarntz** [comment]
@joemcgill 
> One thing I'm still unclear about is the criteria that "Measurement should be taken for the current commit as well as the previous commit for comparison".

I think that's not necessary for the MVP. I believe the original thought behind this was that, eventually, it would be great to have more proactive reporting or alerting, e.g. if performance regressed more than x% in comparison with another set of results. However, even for that, we have not yet discussed what that comparison should be based on (previous commit, previous release, ...?), so I think we can ignore this for now. This can be explored some time after the MVP, once we feel confident about the accuracy of the performance metrics we're collecting.

**Comment by mukeshpanchal27** [comment]
Thanks @joemcgill, The AC look good to me.

> @mukeshpanchal27 it looks to me like those tests are failing due to file changes occurring during the test run. You'll likely need to add a rule to the `.gitignore` file to make sure any files generated during tests are not picked up by git.

@joemcgill I have added those files in `.gitignore` but still the E2E tests going fails.

> @mukeshpanchal27 I've updated the AC on this one to hopefully clarify the requirements a bit, based on conversations from #1. Let me know if you have any questions.

@joemcgill Do we needs to add bunch of issues based of your Requirements list or we will work in single PR?

**Comment by joemcgill** [comment]
It looks like you've already fixed the E2E test issues you were working on, but if you not and you need a second pair of eyes, let me know. I think it's ok for us to track everything in the AC to this issue if that's ok with you.

I'd suggest that we create a feature branch in this repo for tracking this work and you can open a PR against that feature branch with work in progress so it's easier to give feedback and review progress. We can break up parts of these AC into smaller PRs that are opened against that feature branch if it's useful to do the work in smaller steps.

**Comment by felixarntz** [comment]
@joemcgill The ACs here look good to me. The main piece to determine early on in the implementation phase will now be to see whether we should use the e2e or cURL approach.

Definitely the first is preferable for the long term, but if we run into major problems or blockers, then it's okay to go with the simpler cURL approach for now, just in the interest of time.

**Comment by joemcgill** [comment]
Thanks @felixarntz. Your comment about implementation approach is exactly in line with my expectations as well.

**Comment by mukeshpanchal27** [comment]
Resolved by #16.
